cpp/src/arrow/array.h

```diff
@@ -325,7 +325,7 @@ class ARROW_EXPORT Array {
     } else {
       null_bitmap_data_ = NULLPTR;
     }
-    data_ = data;
+    data_ = data->Copy();
```

Comment:
Any reason to explicit it?

Reply:
`Copy()` produces a deep copy, whereas `data_ = data` would mean that subsequently modifying the argument also modifies the array:

```
shared_ptr<ArrayData> somedata = ArrayData::Make(int32(), 0);
Int32Array arr(somedata);
somedata->length = 3;
arr->length(); // 3
```

Reply:
I had forgotten that data_ is a pointer and not the ArrayData object directly.1

Reply:
Are You sure that it is the expected behaviour? If so We should test it.
cc @pitrou

Reply:
This doesn't look right to me